Add names next to scores in FFA in /match and sort by score.

mrcookieunderscore13 opened the pull request:
fixes https://github.com/PGMDev/PGM/issues/619

Typing `/match` on FFA now looks like this:
<img width="481" alt="Screen Shot 2020-08-19 at 3 05 41 PM" src="https://user-images.githubusercontent.com/28692086/90684313-9b972100-e22d-11ea-9a18-1f59265c662b.png">

Signed-off-by: mrcookie

mrcookieunderscore13 committed:
Add names next to scores in FFA in /match and sort by score.

Signed-off-by: mrcookie <athertonmook@gmail.com>

Pablete1234 reviewed (changes requested):
If you're using text components, don't concatenate TextColors/ChatColors as strings, there are a couple places where you're doing stuff like `ChatColor.DARK_AQUA + "Score: "`, that should go away.

Also make use of a `TextComponent.builder`, it will simplify the code alot

applenick commented:
Nice job!  I agree with what @Pablete1234 suggested

Also regarding the formatting, would be better to perhaps differentiate the player from the score by color. Perhaps making the score white. Also making the `:` gray would improve the look too.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +86,47 @@ public double getScore(Competitor competitor) {
   /** Gets the status message for the match. */
-  public String getStatusMessage() {
-    StringBuilder message = new StringBuilder(this.getScoreMessage());
+  public TextComponent getStatusMessage() {
+    TextComponent message = this.getScoreMessage();
```
```suggestion
  public Component getStatusMessage() {
    Component message = this.getScoreMessage();
```

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +86,47 @@ public double getScore(Competitor competitor) {
+                              .append(
+                                  TextComponent.of(
+                                      x.getKey().getColor().toString()
+                                          + ((int) (double) x.getValue())))));
```
`(int) x.getValue().doubleValue()`

same for the `scorePair.getValue` call

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +86,47 @@ public double getScore(Competitor competitor) {
-    List<String> scores = Lists.newArrayList();
-    for (Entry<Competitor, Double> scorePair : this.scores.entrySet()) {
-      scores.add(scorePair.getKey().getColor().toString() + ((int) (double) scorePair.getValue()));
+  public TextComponent getScoreMessage() {
+    List<TextComponent> scoreMessages = Lists.newArrayList();
+    FreeForAllMatchModule ffamm = match.getModule(FreeForAllMatchModule.class);
```
```suggestion
  public Component getScoreMessage() {
    final List<Component> scoreMessages = Lists.newArrayList();
    final FreeForAllMatchModule ffamm = match.getModule(FreeForAllMatchModule.class);
```

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -2,14 +2,10 @@
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Remove this wildcard import

Electroid commented:
Maybe we should only show the top 10 scores? For a 50 player FFA, `/match` will become unreadable

Pablete1234 commented:
> Maybe we should only show the top 10 scores? For a 50 player FFA, `/match` will become unreadable

Possibly show top 9 + yours, so you can always check your own scores

mrcookieunderscore13 commented:
I limited the list to 10 entires, fixed the concatenation issues, and changed the colors around as to AppleNick's comment. For pablete's suggestion to add your own score, I think the best way is to add another line to the `/match` command that shows your score--rather than sticking it to the end of the list. Maybe `You: <score>`. Thoughts?

Pablete1234 commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +90,50 @@ public double getScore(Competitor competitor) {
+                          .append(TextComponent.of((int) x.getValue().doubleValue()))
+                          .color(TextColor.WHITE)
+                          .build()));
+      while (scoreMessages.size() > 10) {
+        scoreMessages.remove(scoreMessages.size() - 1);
+      }
```
use a `.limit(10)` on the stream instead

Pablete1234 commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +90,50 @@ public double getScore(Competitor competitor) {
+    if (ffamm != null) {
+      this.scores.entrySet().stream()
+          .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
+          .forEachOrdered(
+              x ->
+                  scoreMessages.add(
+                      TextComponent.builder()
+                          .append(x.getKey().getName(NameStyle.VERBOSE))
+                          .append(
+                              TextComponent.builder().append(": ").color(TextColor.GRAY).build())
+                          .append(TextComponent.of((int) x.getValue().doubleValue()))
+                          .color(TextColor.WHITE)
+                          .build()));
```
Use a proper reducer instead of forEachOrdered

KingOfSquares replied:
When working with `TextComponent.of` you can use the color in the primary constructor. 
What I mean is instead of 
`TextComponent.of("Text").color(TextColor.WHITE);`
you do 
`TextComponent.of("Text", TextColor.WHITE);`

Also, on line 105, why are you using a builder instead of just a simple `TextComponent.of()`?

Pablete1234 commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +90,50 @@ public double getScore(Competitor competitor) {
-          .append(this.config.scoreLimit)
-          .append("]");
+          .append(TextComponent.of("  "))
+          .append(TextComponent.of(TextColor.GRAY + "["))
```
Don't use chat colors as strings, use them on the components instead

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +90,50 @@ public double getScore(Competitor competitor) {
+      for (Entry<Competitor, Double> scorePair : this.scores.entrySet()) {
+        scoreMessages.add(
+            TextComponent.of(
+                scorePair.getKey().getColor().toString()
```
Use `TextFormatter#convert` on the color to use it as a TextColor instead.

KingOfSquares commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -88,26 +90,50 @@ public double getScore(Competitor competitor) {
+          .append(TextComponent.of("  "))
+          .append(TextComponent.of(TextColor.GRAY + "["))
+          .append(TextComponent.of(this.config.scoreLimit))
+          .append(TextComponent.of("]"));
```
Any reason why you are splitting up these symbols into different `append`s? Seems like they could be a single one(since you are not doing any newlines)

mrcookieunderscore13 commented:
This is what the message looks like now:
<img width="490" alt="Screen Shot 2020-08-23 at 2 03 54 PM" src="https://user-images.githubusercontent.com/28692086/91742342-c1042300-eb7b-11ea-9af2-faeab7aee6cb.png">

Pablete1234 commented on core/src/main/java/tc/oc/pgm/score/ScoreMatchModule.java:
```diff
@@ -92,26 +95,60 @@ public double getScore(Competitor competitor) {
-    for (Entry<Competitor, Double> scorePair : this.scores.entrySet()) {
-      scores.add(scorePair.getKey().getColor().toString() + ((int) (double) scorePair.getValue()));
+  public Component getScoreMessage(MatchPlayer matchPlayer) {
+    List<Component> scoreMessages = Lists.newArrayList();
+    final FreeForAllMatchModule ffamm = match.getModule(FreeForAllMatchModule.class);
+    if (ffamm != null) {
+      scoreMessages =
+          this.scores.entrySet().stream()
+              .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
+              .limit(10)
+              .map(
+                  x ->
+                      TextComponent.builder()
+                          .append(x.getKey().getName(NameStyle.VERBOSE))
+                          .append(TextComponent.of(": ", TextColor.GRAY))
+                          .append(TextComponent.of((int) x.getValue().doubleValue()))
+                          .color(TextColor.WHITE)
+                          .build())
+              .collect(Collectors.toList());
+    } else {
+
+      for (Entry<Competitor, Double> scorePair : this.scores.entrySet()) {
+        scoreMessages.add(
+            TextComponent.of(
+                ((int) scorePair.getValue().doubleValue()),
+                TextFormatter.convert(scorePair.getKey().getColor())));
+      }
     }
```
You can simplify this using a reducer instead of a collector:
```java
    TextComponent result =
        this.scores.entrySet().stream()
            .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
            .limit(10)
            .reduce(
                TextComponent.builder(),
                (builder, entry) ->
                    builder
                        .append(entry.getKey().getName(NameStyle.VERBOSE))
                        .append(": ", TextColor.GRAY)
                        .append("" + entry.getValue().intValue(), TextColor.WHITE),
                ComponentBuilder::append)
            .build();
```

mrcookieunderscore13 replied:
Well, I need the output to be a list, so I can use `TextFormatter.list(...)` to list the names in a nice format: `Player1, Player2, ... , and PlayerN`. Using a collector seems like the proper way right?

Pablete1234 replied:
A joining collector may be a better alternative then, look at how the `Collectors.joining` collector works for strings and do a similar one for this

mrcookieunderscore13 replied:
Using Collectors.joining, there is no easy way to have the last element separated like `<element n-1>, and <element n>`, rather than just a comma. I don't understand what is wrong with the method I am currently using.

KingOfSquares replied:
@mrcookieunderscore13 Maybe you are looking for something like what is in ```TextFormatter#list``` ?

mrcookieunderscore13 replied:
@KingOfSquares I do use that method. I use the stream to make an ordered list of the players, and then a few lines down use `TextFormatter#list` to format it properly.

KingOfSquares replied:
Right, then I think I misunderstood the argument between you and Pablo

applenick commented:
@mrcookieunderscore13 
Related to the new screenshot, format looks great! 
Perhaps you could make `You:` the same color as the player’s name (as in if I was participating and my name was dark_red, the `You` would match)? Also consider making `Score` and `You` translatable.

mrcookieunderscore13 commented:
Applenick I made your changes, but the computer I dev on died, so thats why I haven't commited the change yet. I should be good to go in about a week.

applenick commented:
@mrcookieunderscore13 Ah thanks for the update!

mrcookieunderscore13 committed:
Added translations

Signed-off-by: mrcookie <athertonmook@gmail.com>

mrcookieunderscore13 committed:
Change color of "You: " to match your team color

Signed-off-by: mrcookie <athertonmook@gmail.com>
